File: ruby/birthdays.rb
#!/usr/bin/env ruby

class Birthday
  include Comparable
  attr_reader :year, :month, :day, :name

  def initialize(year, month, day, name)
    return nil unless month >= 1 and month <= 12 and
                      day >= 1 and day <= 31 and
                      name != nil and !name.empty?
    @year = year
    if @year < 1000 then
      @year = 0
    end
    @month = month
    @day = day
    @name = name.strip
  end

  def Birthday.newFromLine(line)
    year, month, day, name = line.split(/[\s\-]+/, 4)
    new(year.to_i, month.to_i, day.to_i, name)
  end

  def stringWithAgeIn(ageIn)
    (@year > 0 ? sprintf("%04u", @year) : "?   " ) +
      sprintf("-%02u-%02u %s", @month, @day, @name) +
      ((@year > 0 and ageIn) ? " (#{(ageIn - @year)})" : "")
  end

  def to_str
    stringWithAgeIn(nil)
  end

  def <=>(another)
    delta = @month <=> another.month
    return delta unless delta == 0
    delta = @day <=> another.day
    return delta unless delta == 0
    delta = @year <=> another.year
    # invert years, because THIS year should be the first
    return -delta unless delta == 0
    return @name <=> another.name
  end
    
end

def addFlatFileBirthdays(list)
  return unless handle = File.new(ENV['HOME'] + '/birthdays', 'r')
  handle.each do |line|
    list.push(Birthday.newFromLine(line))
  end
  handle.close
end

def addOSXBirthdays(list)
  begin
    require 'osx/cocoa'
  rescue LoadError => e
    return
  end

  OSX.require_framework "AddressBook"

  ab = OSX::ABAddressBook.sharedAddressBook

  ab.people.to_a.each do |p|
    nameString = ""
    first = p.valueForProperty(OSX::KABFirstNameProperty).to_s
    if (!first.empty?) then
      nameString += first
    end
    nickname = p.valueForProperty(OSX::KABNicknameProperty).to_s
    if (!nickname.empty?) then
      if (!first.empty?)
        nameString += " ‘#{nickname}’"
      else
        nameString += nickname
      end
    end
    if (!nameString.empty?) then
      nameString += " "
    end
    last = p.valueForProperty(OSX::KABLastNameProperty).to_s
    if !last.empty? then
      nameString += last
    end
    company = p.valueForProperty(OSX::KABOrganizationProperty).to_s
    if !company.empty? then
      if (nameString.empty?) then
        nameString = company
      else
        # is kABPersonFlags really an int?
        if (flags = p.valueForProperty(OSX::KABPersonFlags)) and
           (flags.intValue & OSX::KABShowAsMask == OSX::KABShowAsCompany) then
          nameString = "#{company}: #{nameString}"
        else
          if nameString[-1,1] != " " then
            nameString += " "
          end
          nameString += "[#{company}]"
        end
      end
    end
    if birth = p.valueForProperty(OSX::KABBirthdayProperty) then
      #birthString = sprintf("%04u", birth.yearOfCommonEra) + "-" +
      #sprintf("%02u", birth.monthOfYear) + "-" +
      #sprintf("%02u", birth.dayOfMonth)
      b = Birthday.new(birth.yearOfCommonEra, birth.monthOfYear,
                       birth.dayOfMonth, nameString)
      list.push(b)
    end
  end
end

l = Array.new
addFlatFileBirthdays(l)
addOSXBirthdays(l)
now = Time.now
today = Birthday.new(now.year, now.month, now.day, "\x1b[1m*** TODAY ***")
l.push(today)
l.sort!

showbefore = 3
showafter = 5
todayindex = l.index(today)

# dates from the year before this year
if (showbefore - todayindex > 0) then
  (showbefore - todayindex).downto(1) do |i| 
    if (l.length - i > todayindex) then
      puts l[l.length - i].stringWithAgeIn(today.year - 1)
    end
  end
  puts "----------"
end

# all dates from this year that are before today
showbefore.downto(1) do |i|
  if (todayindex - i >= 0) then
    puts l[todayindex - i].stringWithAgeIn(today.year)
  end 
end

# reverse video TODAY
puts "\x1b[7m" + today + "\x1b[m"

# following dates this year
1.upto(showafter) do |i|
  if (todayindex + i < l.length) then
    puts l[todayindex + i].stringWithAgeIn(today.year)
  end
end

# dates next year
if (todayindex + showafter >= l.length) then
  puts "----------"
  0.upto(todayindex + showafter - l.length - 1) do |i|
    if (i < todayindex) then
      puts l[i].stringWithAgeIn(today.year + 1)
    end
  end
end

#ab.save

File: cpp/birthdays.cpp
#include <Core/Core.h>

class Birthday : Comparable <Birthday, Moveable <Birthday> > {
public:
	String name;
	int year;
	int month;
	int day;
	Birthday();
	Birthday(const String&);
	Birthday(const Date&, const String&);
	void Print(int);
	int Compare(const Birthday&) const;
	bool operator<(const Birthday&);
	bool operator>(const Birthday&);
	bool operator==(const Birthday&);
	
};

Birthday::Birthday(const Date &date, const String &who) {
	year = date.year;
	month = date.month;
	day = date.day;
	name = who;
}

Birthday::Birthday(const String &line) {
	Vector <String> tokens = Split(line, ' ', true);
	static Value v;
	switch (tokens.GetCount()) {
		default:
		for (int i = 3; i < tokens.GetCount(); i++)
			name += (i > 3 ? " " : "") + tokens[i];
		case 3:
		if ((v = StdConvertIntNotNull().Scan(tokens[2])) &&
			!IsError(v) && (int)v > 0 && (int)v <= 31)
			day = v;
		else day = -1;
		case 2:
		if ((v = StdConvertIntNotNull().Scan(tokens[1])) &&
			!IsError(v) && (int)v > 0 && (int)v <= 12)
			month = v;
		else month = -1;
		case 1:
		if ((v = StdConvertIntNotNull().Scan(tokens[0])) &&
			!IsError(v) && (int)v >= 1900 && (int)v <= 9999)
			year = v;
		else year = -1;
		case 0:
		break;
	}
}
	
void Birthday::Print(int y = -1) {
	Cout() << (year > 0 ? Format("%04d", year) : "?   ") << "-"
	       << (month > 0 ? Format("%02d", month) : "? ") << "-"
	       << (day > 0 ? Format("%02d", day) : "? ") << ": " << name;
	if (y > 0 && year > 1)
		Cout() << " (" << (y - year) << ")";
	Cout() << '\n';
}

int Birthday::Compare(const Birthday &other) const {
	int delta = month - other.month;
	if (delta == 0) {
		delta = day - other.day;
		if (delta == 0) {
			delta == year - other.year;
			if (delta == 0) {
				return name.Compare(other.name);
			} else {
				return -delta;
			}
		} else return delta;
	} else return delta;
}

inline bool Birthday::operator<(const Birthday &other) {
	return (Compare(other) < 0);
}

inline bool Birthday::operator>(const Birthday &other) {
	return (Compare(other) > 0);
}

inline bool Birthday::operator==(const Birthday &other) {
	return (Compare(other) == 0);
}

CONSOLE_APP_MAIN {
	Vector<Birthday> list;

	FileIn birthdays(String(getenv("HOME")) + "/birthdays");
	while (!birthdays.IsEof()) list << Birthday(birthdays.GetLine());
	birthdays.Close();

	Birthday today = Birthday(Date(), "\x1b[1m*** TODAY ***");
	time_t now = time(NULL);
	struct tm *bla = localtime(&now);
	today.year = bla->tm_year + 1900;
	today.month = bla->tm_mon + 1;
	today.day = bla->tm_mday;
	list << today;

	Sort(list);
	
	int todayindex = FindBinary(list, today);
	int showbefore = 3;
	int showafter = 5;
	
	// dates from the year before this year
	if (showbefore - todayindex > 0) {
		for (int i = showbefore - todayindex; i > 0; i--)
			// only show if not longer than a year ago
			if (list.GetCount() - i > todayindex)
				list[list.GetCount() - i].Print(today.year - 1);
		Cout() << "----------" << '\n';
	}

	// all dates from this year that are before today
	for (int i = showbefore; i > 0; i--)
		if (todayindex - i >= 0)
			list[todayindex - i].Print(today.year);
	
	// ‘reverse video’
	Cout() << "\x1b[7m";
	today.Print();
	Cout() << "\x1b[m";
	
	// following dates this year
	for (int i = 1; i <= showafter; i++) 
		if (todayindex + i < list.GetCount())
			list[todayindex + i].Print(today.year);
	
	// dates next year
	if (todayindex + showafter >= list.GetCount()) {
		Cout() << "----------" << '\n';
		for (int i = 0; i < todayindex + showafter - list.GetCount(); i++)
			if (i < todayindex)
				list[i].Print(today.year + 1);
	}
}
